src/main/java/it/pagopa/pm/gateway/controller/XPayPaymentController.java

```diff
@@ -0,0 +1,232 @@
+
+    @Value("${pgs.xpay.request.responseUrl}")
+    private String XPAY_RESPONSE_URL;
+
+    @Value("${pgs.xpay.apiKey}")
+    private String API_KEY;
+
+    @Value("${pgs.xpay.secretKey}")
+    private String SECRET_KEY;
+
+    @Autowired
+    private PaymentRequestRepository paymentRequestRepository;
+
+    @Autowired
+    private PaymentResponseRepository paymentResponseRepository;
+
+    @Autowired
+    private XpayService service;
+
+    @PostMapping()
+    public ResponseEntity<XPayAuthResponse> requestPaymentsXPay(@RequestHeader(value = X_CLIENT_ID) String clientId,
+                                                                @RequestHeader(required = false, value = MDC_FIELDS) String mdcFields,
+                                                                @RequestBody XPayAuthRequest pgsRequest) {
+
+        log.info("START POST - " + REQUEST_PAYMENTS_XPAY);
+        setMdcFields(mdcFields);
+        String idTransaction = pgsRequest.getIdTransaction();
+
+        if (ObjectUtils.anyNull(pgsRequest) || pgsRequest.getGrandTotal().equals(BigInteger.ZERO)) {
+            log.error("Bad Request - " + BAD_REQUEST_MSG);
+            return createxPayAuthResponse(BAD_REQUEST_MSG, HttpStatus.BAD_REQUEST, null);
+        }
+
+        if (!VALID_CLIENT_ID.contains(clientId)) {
+            log.error(format("Client id %s is not valid", clientId));
+            return createxPayAuthResponse(BAD_REQUEST_MSG_CLIENT_ID, HttpStatus.BAD_REQUEST, null);
+        }
+
+        if(Objects.nonNull(paymentRequestRepository.findByIdTransaction(idTransaction))) {
+            log.warn("Transaction " + idTransaction + " has already been processed previously");
+            return createxPayAuthResponse(TRANSACTION_ALREADY_PROCESSED_MSG, HttpStatus.UNAUTHORIZED, null);
+        }
+        log.info("END POST - " + REQUEST_PAYMENTS_XPAY);
+        return createAuthPaymentXPay(pgsRequest, clientId, mdcFields);
+    }
+
+    private ResponseEntity<XPayAuthResponse> createxPayAuthResponse(String errorMessage, HttpStatus status, String requestId) {
+        XPayAuthResponse respone = new XPayAuthResponse();
+        respone.setRequestId(requestId);
+        if(StringUtils.isEmpty(errorMessage)) {
+            String urlRedirect = StringUtils.join(PGS_RESPONSE_URL_REDIRECT, requestId);
+            respone.setUrlRedirect(urlRedirect);
+        } else {
+            respone.setError(errorMessage);
+            log.info("END POST - " + REQUEST_PAYMENTS_XPAY);
+        }
+        return ResponseEntity.status(status).body(respone);
+    }
+
+    private ResponseEntity<XPayAuthResponse> createAuthPaymentXPay(XPayAuthRequest pgsRequest, String clientId, String mdcFields) {
+       log.info("START - XPay Request Payment Authorization ");
+        AuthPaymentXPayRequest xPayRequest = createXPayRequest(pgsRequest);
+
+        PaymentRequestEntity paymentRequestEntity = new PaymentRequestEntity();
+        String transactionId = pgsRequest.getIdTransaction();
+        try{
```

Comment:
@DavideAmatoNexi You can merge these two try/catch clauses and also move generateRequestEntity() method above as soon as you create the PaymentRequestEntity object:

`PaymentRequestEntity paymentRequestEntity = generateRequestEntity(clientId, mdcFields, transactionId, paymentRequestEntity, xPayRequest.getTimeStamp());`